### ISSUE TITLE
[Feature Request] - KNeighborsThresholdClassifier

### ISSUE THREAD

**Issue by jolespin** [jolespin/skclust#42]
# KNeighborsThresholdClassifier

## Overview

A distance-weighted KNN classifier with per-class reject option and
margin-based confidence reporting. Goes in `skclust/neighbors.py`.

The class wraps sklearn's `KNeighborsClassifier` and `NearestNeighbors` but
hides the precomputed-graph optimization behind a clean API: users pass raw
feature matrices, the class handles neighbor indexing internally.

The distinguishing feature: per-class probability thresholds calibrated via
one-vs-rest F1 maximization, optionally under user-supplied perturbation. This
enables a reject option — predictions below the class-specific threshold are
rejected rather than assigned.

Default metric is cosine. Default weighting is distance (inverse distance).

---

## File location

```
skclust/
    __init__.py          # add import
    neighbors.py         #  class lives here
```

---

## Dependencies

```python
import numpy as np
import pandas as pd
import scipy.sparse as sps
from loguru import logger
from sklearn.neighbors import (
    KNeighborsClassifier,
    NearestNeighbors,
)
```

No additional dependencies beyond what skclust already uses.

---

## Class signature

```python
class KNeighborsThresholdClassifier:
    """
    Distance-weighted KNN classifier with per-class reject option
    and margin-based confidence reporting.

    Parameters
    ----------
    n_neighbors : int, default=5
        Number of neighbors for classification.
    metric : str, default="cosine"
        Distance metric for neighbor search. Passed to
        sklearn.neighbors.NearestNeighbors.
    weights : str, default="distance"
        Weight function. "distance" uses inverse distance;
        "uniform" weights all neighbors equally.
    n_jobs : int, default=-1
        Parallel jobs for neighbor search. -1 uses all cores.
    calibration_batch_size : int, default=8192
        Batch size for processing calibration queries. Controls
        memory usage during calibrate(). Not an algorithmic
        parameter — adjust based on available RAM.
    """
```

---

## Attributes (set by fit)

```
classes_          : np.ndarray  — sorted unique classes from training y
n_classes_        : int         — len(classes_)
n_features_in_    : int         — number of features in training X
is_calibrated     : bool        — True after calibrate() has been called
class_to_threshold : pd.Series  — per-class threshold (index=classes_), None before calibration
fallback_threshold_ : float     — median threshold used for small classes, None before calibration
```

---

## Internal state (set by fit, not part of public API)

```
_nn_index         : NearestNeighbors — fitted neighbor index on training X
_knn              : KNeighborsClassifier — fitted on precomputed identity self-graph
_graph_fit        : sps.csr_matrix — identity self-graph (n_train, n_train) * 1e-6
_X_fit            : np.ndarray — stored training features (needed for LOO calibration)
_y_fit            : np.ndarray — stored training labels
_y_codes          : np.ndarray (int32) — integer-coded training labels
_code_to_threshold : np.ndarray — float64 threshold array indexed by class code
```

---

### `fit(self, X, y) -> self`

1. Convert X to np.ndarray if DataFrame (store `.values`). Store as `_X_fit`.
2. **Validate metric:** if `metric == "precomputed"`, raise
   `ValueError("metric='precomputed' is not supported. Pass raw features; "
   "the precomputed neighbor graph is handled internally.")`.
   The class uses `metric="precomputed"` on the internal KNeighborsClassifier
   as an optimization — but the user always passes raw feature matrices, and
   the class stores them to compute query-to-reference distances at predict
   time. Exposing `metric="precomputed"` would break predict() for
   out-of-sample data since the original features would not be stored.
3. Store y as `_y_fit`. Compute `classes_` via `np.sort(np.unique(y))`.
3. Encode y to integer codes via `pd.Categorical(y, categories=classes_).codes`.
4. Build `NearestNeighbors(n_neighbors=n_neighbors, metric=metric, n_jobs=n_jobs).fit(X)`.
5. Build identity self-graph: `sps.eye(n_train, format="csr") * 1e-6`.
   - **Why:** `KNeighborsClassifier(metric="precomputed")` requires a square
     (n_train, n_train) matrix at fit time but never reads its values —
     predictions come entirely from the query graph. Every point IS its own
     nearest neighbor at distance ~0, so this is the exact 1-NN self-graph.
6. Fit `KNeighborsClassifier(n_neighbors=n_neighbors, metric="precomputed", weights=weights)` on the self-graph + y.
7. Set `is_calibrated = False`, `class_to_threshold = None`.
8. Log: `"Fitted on {n_train} samples / {n_classes_} classes"`.
9. Return self.

### `calibrate(self, X=None, y=None, perturbation_fn=None, n_replicates=10, n_per_class=None, min_positives=20, n_grid=101, random_state=None) -> self`

**Purpose:** learn a per-class probability threshold that maximizes per-class
F1, optionally under perturbation.

**Parameters:**

```
X               : np.ndarray or None
                  Calibration features. None = use training data (LOO mode).
y               : np.ndarray or None
                  Calibration labels. None = use training labels.
                  Ignored when X is None.
perturbation_fn : callable or None
                  Signature: (X_batch, y_batch, rng) -> X_perturbed
                  X_batch: np.ndarray (batch_size, n_features)
                  y_batch: np.ndarray (batch_size,) — original labels for
                           the batch, enabling per-sample perturbation
                           (e.g., metadata lookup). Can be ignored.
                  rng: np.random.Generator
                  Returns: np.ndarray, same shape as X_batch
                  None = calibrate on clean data (no perturbation).
n_replicates    : int — number of independent perturbation draws per sample
n_per_class     : int or None — max calibration samples per class for balance.
                  None = use all available samples. Recommended when class
                  sizes vary by more than ~10x.
min_positives   : int — classes with fewer calibration positives use fallback
n_grid          : int — threshold grid resolution (0.0 to 1.0 inclusive)
random_state    : int or None — seed for RNG (balance sampling + perturbation).
                  None = non-deterministic (sklearn convention).
```

Batch size for calibration queries is set via `calibration_batch_size` in
the constructor, not here.

**Procedure:**

1. Validate that fit() has been called (check `_nn_index` exists).
2. Determine mode:
   - **LOO mode** (X is None): use `_X_fit` and `_y_fit`. Set
     `leave_one_out = True`.
   - **External mode** (X provided): use X and y. Set
     `leave_one_out = False`.
3. Balance-sample calibration queries: if `n_per_class` is not None, for each
   class sample `min(class_size, n_per_class)` samples. If `n_per_class` is
   None, use all samples. Store their indices (into the calibration array)
   and, for LOO mode, their positions in the training reference set (needed
   for self-exclusion).
4. Loop: for each replicate (0 to n_replicates-1):
   - For each batch of calibration queries (sized by `self.calibration_batch_size`):
     a. If `perturbation_fn` is not None, call it:
        `X_perturbed = perturbation_fn(X_batch, y_batch, rng)`
        Otherwise: `X_perturbed = X_batch` (copy not needed, read-only).
     b. Build neighbor graph via `_neighbor_graph()`:
        - If LOO mode: pass `self_positions` (each query's row index in the
          training reference set) to exclude self-matches.
        - If external mode: pass `self_positions=None`.
     c. Get probabilities from `_knn.predict_proba(graph)`.
     d. Collect `argmax` (predicted codes) and `max` (max probability).
   - Log: `"Replicate {i+1}/{n_replicates} complete"`.
5. Pool all replicates: tile true codes n_replicates times, concatenate
   predicted codes and max probabilities.
6. Call `_calibrate_thresholds()` with pooled arrays.
7. Store `_code_to_threshold`, `fallback_threshold_`, build
   `class_to_threshold` Series, set `is_calibrated = True`.
8. Log: `"Calibrated {n}/{n_classes_} classes directly; fallback={f:.3f}"`.
9. Return self.

### `predict(self, X, index=None) -> pd.DataFrame`

**Parameters:**

```
X     : np.ndarray — query features (n_query, n_features)
index : array-like or None — row index for the returned DataFrame
```

**Procedure:**

1. Validate fit. Convert X to ndarray if needed.
2. Build neighbor graph via `_neighbor_graph(X)` (no self-exclusion).
3. Get probabilities: `_knn.predict_proba(graph)` → (n_query, n_classes).
4. Compute:
   - `predicted_codes = probabilities.argmax(axis=1)`
   - `top1_prob = probabilities.max(axis=1)`
   - `top2_codes = np.argpartition(probabilities, -2, axis=1)[:, -2]`
   - `top2_prob = probabilities[np.arange(n), top2_codes]`
   - `margin = top1_prob - top2_prob`
5. Build DataFrame (index=index):
   - `"predicted_class"` → `classes_[predicted_codes]`
   - `"probability"` → top1_prob
   - `"runner_up_class"` → `classes_[top2_codes]`
   - `"runner_up_probability"` → top2_prob
   - `"margin"` → margin
6. If `is_calibrated`:
   - `"threshold"` → `_code_to_threshold[predicted_codes]`
   - `"passed"` → probability >= threshold
   - `"assigned_class"` → predicted_class where passed, else NaN
7. Return DataFrame.

### `predict_proba(self, X) -> np.ndarray`

1. Build neighbor graph, return `_knn.predict_proba(graph)`.
2. Shape: (n_query, n_classes_). Column order follows `classes_`.

### `score(self, X, y) -> float`

1. If calibrated: rejection-aware macro-F1 (below-threshold predictions
   cost recall but add no false positive — see `_score_with_rejection`).
2. If not calibrated: standard macro-F1 via sklearn.

---

### `_neighbor_graph(self, X_query, self_positions=None) -> sps.csr_matrix`

Sparse (n_query, n_reference) distance graph.

1. Call `_nn_index.kneighbors(X_query, n_neighbors=n_neighbors + (self_positions is not None))`.
2. If `self_positions` is not None:
   - Build boolean mask: `is_self = indices == self_positions[:, None]`
   - Handle case where self not found in neighbors (perturbed genome moved
     far enough that it's not in top-k+1): set `is_self[~is_self.any(axis=1), -1] = True`
     to drop the farthest neighbor instead.
   - Remove self column: reshape distances and indices to (n_query, n_neighbors).
3. Build CSR matrix:
   ```python
   sps.csr_matrix(
       (distances.ravel(), indices.ravel(),
        np.arange(0, distances.size + 1, n_neighbors)),
       shape=(n_query, self._nn_index.n_samples_fit_),
   )
   ```

### `_calibrate_thresholds(self, true_codes, predicted_codes, max_probability) -> tuple[np.ndarray, float]`

Per-class F1-maximizing threshold on a grid.

1. Create grid: `np.linspace(0.0, 1.0, n_grid)`.
2. For each class code:
   - Skip if fewer than `min_positives` true positives.
   - For each threshold in grid, compute TP, FP, FN for the one-vs-rest
     binary problem: a prediction counts only if predicted_code == this class
     AND max_probability >= threshold.
   - Compute F1 at each grid point.
   - **Ties break high:** use `grid[len(f1) - 1 - int(np.argmax(f1[::-1]))]`.
     This selects the HIGHEST threshold among ties. Rationale: if F1 is flat
     at the bottom of the grid, it means no calibration sample had probability
     that low — the flat region is an artifact, not evidence that rejection
     is unnecessary.
3. Classes that were skipped get NaN.
4. Compute fallback: `np.nanmedian(code_to_threshold)`.
5. Fill NaN with fallback: `np.where(np.isnan(...), fallback, ...)`.
6. Return `(code_to_threshold, fallback)`.

### `_score_with_rejection(self, true_codes, predicted_codes, max_probability, labels) -> float`

1. `passed = max_probability >= _code_to_threshold[predicted_codes]`
2. Among passed predictions, compute per-class TP, FP, FN via `np.bincount`.
3. FN includes both misclassifications AND rejections (rejected predictions
   cost a false negative for their true class).
4. Compute F1 per class, return macro average over `labels`.

---

## Edge cases to handle

1. **predict() before fit():** raise `ValueError("Not fitted. Call fit() first.")`.
2. **calibrate() before fit():** same error.
3. **predict() before calibrate():** works fine — return DataFrame without
   threshold/passed/assigned_class columns.
4. **n_per_class > class size:** use all members of that class (take
   `min(class_size, n_per_class)`).
5. **Class absent from calibration set:** (e.g., all predictions were wrong,
   fewer than min_positives positives) → fallback threshold.
6. **perturbation_fn is None:** calibrate on clean data. Each replicate is
   identical, so effectively n_replicates=1 in terms of information. Log a
   warning: `"No perturbation_fn; replicates are identical. Consider n_replicates=1."`.
7. **X is DataFrame:** convert to `.values` silently. Do NOT store feature
   names (avoids sklearn feature-name warnings).
8. **y contains classes not seen in fit:** raise ValueError at calibrate/score.
9. **Single-class edge:** technically works but degenerate. Don't special-case.
10. **metric="precomputed":** raise ValueError at `__init__` or `fit`. The
    class uses precomputed graphs internally as an optimization — the user
    should never set this. Out-of-sample prediction requires stored training
    features to compute query-to-reference distances.

---

### Minimal (no calibration)

```python
from skclust.neighbors import KNeighborsThresholdClassifier

clf = KNeighborsThresholdClassifier(n_neighbors=5)
clf.fit(X_train, y_train)
df_predictions = clf.predict(X_test)
# Columns: predicted_class, probability, runner_up_class,
#          runner_up_probability, margin
```

### With calibration (no perturbation)

```python
clf = KNeighborsThresholdClassifier(n_neighbors=5)
clf.fit(X_train, y_train)
clf.calibrate()  # LOO on training data, clean
df_predictions = clf.predict(X_test)
# Additional columns: threshold, passed, assigned_class
```

### With perturbation (domain-specific noise)

```python
def perturbation_fn(X_batch, y_batch, rng):
    """Mask 20% of features to simulate missing data."""
    X_perturbed = X_batch.copy()
    mask = rng.random(X_perturbed.shape) < 0.20
    X_perturbed[mask] = 0.0
    return X_perturbed

clf = KNeighborsThresholdClassifier(n_neighbors=5)
clf.fit(X_train, y_train)
clf.calibrate(perturbation_fn=perturbation_fn, n_replicates=10)
df_predictions = clf.predict(X_test, index=sample_ids)
```

### External calibration set

```python
clf = KNeighborsThresholdClassifier(n_neighbors=5)
clf.fit(X_train, y_train)
clf.calibrate(X=X_val, y=y_val, perturbation_fn=perturbation_fn)
```

### Scoring

```python
# Rejection-aware macro-F1
score = clf.score(X_test, y_test)

# Raw probability matrix for custom evaluation
probabilities = clf.predict_proba(X_test)
```

---

## What this class does NOT do

- **k selection.** The user chooses n_neighbors. Grid search over k is a
  one-liner with sklearn or a for loop — no need to bake it into the class.
- **Embedding or preprocessing.** The class operates on feature matrices.
  Any upstream transformation (SimCLR encoding, L2 normalization, PCA) is
  the user's responsibility before calling fit/predict.
- **Domain-specific perturbation.** Perturbation logic is injected via
  callback. The class handles batching, replication, and threshold math.
- **Custom serialization.** The class is pickle/joblib-compatible by default.
  Wrapping it in gzip or adding save/load is the user's choice.

---

### Why is metric="precomputed" blocked?

To predict on a new sample, the classifier needs to compute the distance
between that sample and every training sample. This requires the raw training
features. Here is what happens in each scenario:

**sklearn's KNeighborsClassifier with metric="precomputed":**
```
fit(distance_matrix)    # (n_train, n_train) — no raw features stored
predict(query_matrix)   # (n_query, n_train) — user must compute this themselves
```
The classifier never sees raw features. The user is fully responsible for
computing the (n_query, n_train) distance matrix before calling predict().
If they lose access to the raw training features, they cannot compute that
matrix, and they cannot predict. The classifier provides no help.

**Our class:**
```
fit(X_train, y_train)   # raw features stored internally
predict(X_query)        # class computes distances automatically
```
The class stores the raw training features at fit time, builds a
NearestNeighbors index on them, and computes query-to-reference distances
internally at predict time. The user never handles distance matrices.

Internally, the class uses metric="precomputed" on the KNeighborsClassifier
as a performance optimization — the NearestNeighbors index computes distances
once, and the KNN classifier receives a precomputed sparse graph. This is an
implementation detail, not a user-facing option.

**Why not accept precomputed + optional raw features?**
If the user provides raw features anyway, they should just set
metric="cosine" (or whatever metric they want) and let the class handle
everything — passing both a distance matrix and raw features adds API
complexity and a consistency invariant (does the matrix match the features?)
for zero benefit. Users with custom distance functions can pass
metric=callable to NearestNeighbors, which already supports this.
